Improve emoji SVG parsing by caching

edwin0cheng opened the pull request:
This PR adds a cache for SVG parsing in the Text Server.

Basically, when we first encounter the document, we parse it as before, but we also note the offsets of other glyphs and store the remaining XML. The next time we see another glyph, we can simply parse that glyph node and insert it back into the stored XML.

Although it is still very slow in the case mentioned in #100278 (which is related to font rendering), the time difference is still significant (on my machine):

Before : 4 mins +
After: 35 secs

Fixes https://github.com/godotengine/godot/issues/100278

edwin0cheng commented:
Seem like Windows build fail due to using designated tantalizers  😞 
Fixed now.

bruvzg commented on modules/text_server_adv/thorvg_svg_in_ot.cpp:
```diff
@@ -91,57 +138,100 @@ FT_Error tvg_svg_in_ot_preset_slot(FT_GlyphSlot p_slot, FT_Bool p_cache, FT_Poin
-					if (vb.size() == 4) {
-						embox_x = vb[2].to_float();
-						embox_y = vb[3].to_float();
+		if (!state->document_map.has(document->svg_document)) {
+			state->document_map[document->svg_document] = TVG_DocumentCache();
+		}
```
```suggestion
```
This check is unnecessary, non-const `[]` operator will automatically insert a new item if it doesn't exist.

edwin0cheng replied:
Done

bruvzg commented on modules/text_server_adv/thorvg_svg_in_ot.cpp:
```diff
@@ -91,57 +138,100 @@ FT_Error tvg_svg_in_ot_preset_slot(FT_GlyphSlot p_slot, FT_Bool p_cache, FT_Poin
+			embox_x = cache.embox_x;
+			embox_y = cache.embox_y;
+
+			DEV_ASSERT(cache.node_caches.has(p_slot->glyph_index));
```
`DEV_ASSERT` will crash, it's too aggressive for emoji parsing, something like **`ERR_FAIL_COND_V(!cache.node_caches.has(p_slot->glyph_index), FT_Err_Invalid_SVG_Document);`** should be used instead.

bruvzg commented on modules/text_server_adv/thorvg_svg_in_ot.cpp:
```diff
@@ -91,57 +138,100 @@ FT_Error tvg_svg_in_ot_preset_slot(FT_GlyphSlot p_slot, FT_Bool p_cache, FT_Poin
+						const String &gl_name = parser->get_named_attribute_value("id");
+						DEV_ASSERT(gl_name.begins_with("glyph"));
+
+						int dot_pos = gl_name.find_char('.');
+						int64_t gl_idx = gl_name.substr(5, (dot_pos > 0) ? dot_pos - 5 : -1).to_int();
+						DEV_ASSERT(p_slot->glyph_index == gl_idx);
 					}
-				}
-				if (parser->has_attribute("width")) {
-					embox_x = parser->get_named_attribute_value("width").to_float();
-				}
-				if (parser->has_attribute("height")) {
-					embox_y = parser->get_named_attribute_value("height").to_float();
+#endif
```
Same, but I do not think these checks are necessary at all.

bruvzg reviewed (commented):
Looks good and works as expected.

The same change should be also applied to `modules/text_server_fb/thorvg_svg_in_ot.*` (exactly same code).

edwin0cheng commented:
And I add fb textserver part too.

edwin0cheng commented:
@bruvzg , do you think these actually fixes #88809 and #100278 ? It still be very slow in #100278 case.

bruvzg commented:
> do you think these actually fixes https://github.com/godotengine/godot/issues/88809 and https://github.com/godotengine/godot/issues/100278 ? It still be very slow in https://github.com/godotengine/godot/issues/100278 case.

Not sure if a lot can be done for pre-rendering, pre-rendering option only exist because it is slow and primarily for MSDF, it should not be used for emojis in a first place.

kiroxas commented on modules/text_server_adv/thorvg_svg_in_ot.cpp:
```diff
@@ -74,6 +74,53 @@ void tvg_svg_in_ot_free(FT_Pointer *p_state) {
+	if (parser->get_node_type() == XMLParser::NODE_ELEMENT) {
+		*p_xml += vformat("<%s", parser->get_node_name());
+
+		bool is_svg_tag = parser->get_node_name() == "svg";
```
There is some small repetitions here, could be  : 
```
if (parser->get_node_type() == XMLParser::NODE_ELEMENT)
{
    *p_xml += vformat("<%s", parser->get_node_name());
    bool is_svg_tag = parser->get_node_name() == "svg";
    for (int i = 0; i < parser->get_attribute_count(); i++) {
		String aname = parser->get_attribute_name(i);
	        if (is_svg_tag && (aname == "viewBox")) {
			PackedStringArray vb = parser->get_named_attribute_value("viewBox").split(" ");

			if (vb.size() == 4) {
				r_embox_x = vb[2].to_float();
				r_embox_y = vb[3].to_float();
			  }
		}
              else if (is_svg_tag && (aname == "width")) {
                        r_embox_x = parser->get_named_attribute_value("width").to_float();
              } else if (is_svg_tag && (aname == "height")) {
                       r_embox_y = parser->get_named_attribute_value("height").to_float();
               } else {
		     *p_xml += vformat(" %s=\"%s\"", aname, parser->get_attribute_value(i));
               }
    }
```  

And as you have the index, you could probably do better than `get_named_attribute_value`

kiroxas commented on modules/text_server_adv/thorvg_svg_in_ot.cpp:
```diff
@@ -75,32 +75,25 @@ void tvg_svg_in_ot_free(FT_Pointer *p_state) {
 			String aname = parser->get_attribute_name(i);
-			if (is_svg_tag && (aname == "viewBox" || aname == "width" || aname == "height")) {
-				continue;
+			if (is_svg_tag && (aname == "viewBox")) {
```
Well, now we can compact even more : 
```
String value = parser->get_attribute_value(i);
(...)
PackedStringArray vb =value.split(" ");
(...)
r_embox_x = value.to_float();
(...)
r_embox_y = value.to_float();
(...)
*p_xml += vformat(" %s=\"%s\"", aname, value);
```

edwin0cheng replied:
Done.

akien-mga commented:
Looks good! Could you squash the commits? See [PR workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) for instructions.

edwin0cheng committed:
Improve emoji SVG parsing by caching

Basically, when we first encounter the document, we parse it as before, but we also note the offsets of other glyphs and store the remaining XML. The next time we see another glyph, we can simply parse that glyph node and insert it back into the stored XML.

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:
